app/models/reports/repository.rb

```diff
@@ -1,64 +1,143 @@
 module Reports
   class Repository
+    include Memery
+    PERCENTAGE_PRECISION = 0
+
     def initialize(regions, periods:, with_exclusions: false)
       @regions = Array(regions)
-      @regions_by_id = @regions.group_by { |r| r.id }
       @with_exclusions = with_exclusions
 
       @periods = if periods.is_a?(Period)
-        Range.new(periods, periods).to_a
+        Range.new(periods, periods)
       else
-        periods.to_a
+        periods
       end
     end
 
-    attr_reader :regions, :periods
-    attr_reader :regions_by_id
+    attr_reader :periods
+    attr_reader :regions
     attr_reader :with_exclusions
 
     delegate :cache, :logger, to: Rails
 
-    def controlled_patients_count
-      cached_query(:controlled_patients_count) do |item|
-        control_rate_query.controlled(item.region, item.period, with_exclusions: with_exclusions).count
+    # Uses Memery to memoize a method, but takes into account our force_cache setting. If force_cache is true,
+    # a caller is asking for all values to be retrieved fresh from the database, so we want to skip memoization.
+    def self.smart_memoize(method)
```

Comment:
I'm not sure I understand this either. If I understand this correctly, when we specify force_cache, the method will not be memoized and we will request the data from the DB each time the method is called (which can be multiple times per request). If we don't specify force_cache, we will memoize and grab from the cache. So it seems like this will improve performance for the case where we use the cache but not for the case where we use the DB. Which seems like it will make the less performant case less performant. Perhaps I'm misunderstanding.

Reply:
The memoization is _just_ for the scope fo the request, for cases where we are in a tight loop and grabbing counts for different regions / periods over and over.  

> If we don't specify force_cache, we will memoize and grab from the cache. So it seems like this will improve performance for the case where we use the cache but not for the case where we use the DB. Which seems like it will make the less performant case less performant. Perhaps I'm misunderstanding.

This is correct, and how it is intended to work.  `force_cache` means to bust the cache and grab fresh values, its only used by us for debugging or when priming the cache.  All caching must be bypassed for that to work, including memoization.

Reply:
side note - probably could use a better name for the `force_cache` piece ...maybe `force_fetch` ? `bust_cache` ?

Reply:
> side note - probably could use a better name for the `force_cache` piece ...maybe `force_fetch` ? `bust_cache` 

see https://github.com/simpledotorg/simple-server/pull/2146